File: Scala/src/main/java/jtypeclasses/printable/domain/Cat.java
package jtypeclasses.printable.domain;

import jtypeclasses.printable.lib.Printable;

public class Cat {

    public final String name;
    public final Integer age;
    public final String color;

    public Cat(String name, Integer age, String color) {
        this.name = name;
        this.age = age;
        this.color = color;
    }

    @Override
    public String toString() {
        return "Cat{" +
                "name='" + name + '\'' +
                ", age=" + age +
                ", color='" + color + '\'' +
                '}';
    }


    public static Printable<Cat> catPrintable = new Printable<Cat>() {
        @Override
        public String format(Cat cat) {
            return cat.name + " is a " + cat.age + " year-old " + cat.color + " cat.";
        }
    };
}

File: Scala/src/main/scala/typeclasses/_01a_my_show/domain/Cat.scala
package typeclasses._01a_my_show.domain

import typeclasses._01a_my_show.lib.Printable
import typeclasses._01a_my_show.lib.PrintableInstances._

final case class Cat(name: String, age: Int, color: String)

object Cat {

  implicit val catPrintable: Printable[Cat] = new Printable[Cat] {
    override def format(cat: Cat): String = {
      val name  = Printable.format(cat.name)
      val age   = Printable.format(cat.age)
      val color = Printable.format(cat.color)
      s"$name is a $age year-old $color cat."
    }
  }
}
